src/Bicep.Core/Analyzers/Linter/Rules/UseProtectedSettingsForCommandToExecuteSecretsRule.cs

```diff
@@ -0,0 +1,108 @@
+using Bicep.Core.Parsing;
+using Bicep.Core.Resources;
+using Bicep.Core.Semantics;
+using Bicep.Core.Semantics.Metadata;
+using Bicep.Core.Semantics.Namespaces;
+using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem;
+using Bicep.Core.Visitors;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Bicep.Core.Analyzers.Linter.Rules
+{
+    public sealed class UseProtectedSettingsForCommandToExecuteSecretsRule : LinterRuleBase
+    {
+        public new const string Code = "use-protectedsettings-for-commandtoexecute-secrets";
+
+        private (string publisher, string type)[] _publisherAndNameList = {
+            // NOTE: This list was obtained by running "az vm extension image list"
+            ("Microsoft.Azure.Extensions", "CustomScript"),
+            ("Microsoft.Compute", "CustomScriptExtension"),
+            ("Microsoft.OSTCExtensions", "CustomScriptForLinux")
+        };
+
+        public (string publisher, string type)[] PublisherAndNameList => _publisherAndNameList;
+
+        public UseProtectedSettingsForCommandToExecuteSecretsRule() : base(
+            code: Code,
+            description: CoreResources.UseProtectedSettingsForCommandToExecuteSecretsRuleDescription,
+            docUri: new Uri($"https://aka.ms/bicep/linter/{Code}")
+        )
+        { }
+
+        public override string FormatMessage(params object[] values)
+            => string.Format(CoreResources.UseProtectedSettingsForCommandToExecuteSecretsRuleMessage, (string)values[0]);
+
+        public override IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel semanticModel)
+        {
+            List<IDiagnostic> diagnostics = new List<IDiagnostic>();
+
+            foreach (ResourceMetadata resource in semanticModel.AllResources)
+            {
+                // We're looking for this pattern:
+                //
+                // resource xxx 'xxx/extensions@yyy' = { // right now, these properties apply to Microsoft.{Hybrid,}Compute/virtualMachines/extensions, but it could be extended
+                //   name: 'xxx'
+                //   properties: {
+                //     publisher: 'Microsoft.Compute'  << === matches entry in PublisherAndNameList
+                //     type: 'CustomScriptExtension' <<=== matches entry in PublisherAndNameList
+                //     autoUpgradeMinorVersion: true
+                //     settings: {  <<== commandToExecute under "settings" instead of "protectedSettings"
+                //       fileUris: split(fileUris, ' ')
+                //       commandToExecute: 'powershell -ExecutionPolicy Unrestricted -File ${firstFileName} ${arguments}'  <<== required
+                //     }
+                //   }
+                // }
+                //
+                // This will be a test failure if commandToExecute contains possible secrets
+
+                var resourceTypeName = resource.TypeReference.FormatType();
```

Comment:
>resourceTypeName [](http://example.com/codeflow?start=20&length=16)

@anthony-c-martin do we have a good way of excluding extensible resource types? Or scoping this down to only `az` resource types?

Reply:
`resource.IsAzResource` should do it.